### DISCUSSION TITLE
General Question or Help about ImGuiFileDialogs

### DISCUSSION THREAD

**Discussion by aiekick** [aiekick/ImGuiFileDialog#40] [Q&A]
Hello,

please use this thread only if you have General questions or need help about ImGuiFileDIalog.

if you have an **issue or need a feature**, please use the issue tab.

**Comment by scottnykl** [comment]
How hard would it be to remove the dependency on Dirent now that C++17 has a std::filesystem library (https://en.cppreference.com/w/cpp/filesystem)? Perhaps it could be an ifdef setting? I like your interface, especially the ability to click on a path in the dialog box and see it transform to editable / copy-able text. I am currently using https://github.com/AirGuanZ/imgui-filebrowser because it remains vanilla C++ / imgui with no other depends, but absolutely love your demo / user interface.

**Reply by aiekick** [reply]
Hello,

thanks for the interest you have for this lib.
yes its planed and im working too on other cool feature. 
but a bit busy for now. 

so could be available for the end of next week maybe

**Reply by aiekick** [reply]
by the way what is the issue to have a dependancy like dirent ? you just need to put the ImGuiFileDialog Directory of the Lib_Only branch in your app and that all, dirent is in this directory directly

**Reply by scottnykl** [reply]
Dirent appears to be functional, and you're right -- I cannot point to any exact issue that prohibits its use. However, moving forward, C++ has std::filesystem and will therefore be directly supported on any platform that has C++ and the std library. With respect to running this software on different OSes as well as different CPU architectures, vanilla C++ and the std library will "just work" (same for cross compiling uProcessors, etc). Dirent would require its devs to port the code to any future system (os/cpu target). I like to keep this in mind as I choose which libraries I integrate into my software packages. 

My main software package has been in use since 2005 and has run on Win,Linux, OS X, Android, Solaris, WebGL, and leveraged processors like intel, ARM, Motorola, Sparc Motorola, RISC V, PowerPC, etc -- so trying to use a smaller library on an OS/CPU target that it hasn't been tested/verified against is something I've come to fear.

I understand if its not worth the effort. What is nice about std::filesystem is it is essentially all value semantics so you won't have to manage any memory like you may need to do when dealing w/ C APIs malloc/free, etc...

**Reply by aiekick** [reply]
i understand, but you know dear imgui have dependency like freetype or stb by ex..
dirent is just a header on win32 and system on unix. 
std::filesystem is not supported on all platfom/compiler or partially [c++17 support](https://en.cppreference.com/w/cpp/compiler_support)
by the way, if we can support std::fielsystem its ok for me. just as a define like you said before.
 std::filesystem appear to be a pain to convert to me with my current filedialog, so if you can, please propose a PR. 
thanks for your understanding.

**Comment by izzynab** [comment]
Hi. I have a problem with including ImGuiFileDialog. Maybe i am just stupid but i dont have any idea how to do it.Here is my stack overflow question:https://stackoverflow.com/questions/67937463/how-to-include-imguifiledialog-with-premake-in-project

**Reply by aiekick** [reply]
Hello,

thanks for using this lib.

btw i cant help you, cause i dont know premake. so its difficult to help.

like its more a general question regarding premake on custom libs.

**Comment by jenergy** [comment]
Hi,
I'm developing a port of Intellivision emulator, with an interface based on Dear Imgui (https://atariage.com/forums/topic/308545-jzintv-imgui/?tab=comments#comment-4833319).
Actually for file/directory chooser:


- On Windows: I developed a custom dll, since I don't like standard folder chooser
- On Linux: I'm using native file dialog extended (https://github.com/btzy/nativefiledialog-extended), I'm not able to compile it on Windows because I'm not using Microsoft environment.
- On Android: I'm looking for a solution :-|

That would be nice to use your solution on Android..and maybe use it also for Windows and Linux too.

Did you try it on Android? That's not reported on main page.
On Android devices I'm able to read/write files from c++, and if needed I can retrieve the root path of internal and external SD.. I think it should be enough to work but it's better to ask you :-)

Thanks

**Comment by aiekick** [comment]
Hello,
i think he will work je dear imgui is working. but not tested.
i think dirent.h is available on android. if confirmed he will work, for directory scanning.
but not sure about the ergonomy, how you control the dialog with multi touch support by ex.
but if you can ensure this part i can add android support in the main page :)

**Reply by jenergy** [reply]
It works! The only thing I noticed is that you cannot switch between internal/external SD. Maybe a kind of "GetDriveList" should be implemented, as for Windows. 
In my application I can surely have that list, by calling Java method. Maybe I can force that list in "DriveList", but I'm not sure that this is possible natively from C++.


![Screenshot_2021-07-14-18-03-03-415_com jzintvimgui](https://user-images.githubusercontent.com/57037540/125655014-7f291564-f1cc-4271-aae4-90b8fd175951.jpg)

**Reply by aiekick** [reply]
nice.
maybe i can define the get drive list as a callback for let user deifne is own
or support android and make some adjustments in that case

**Reply by jenergy** [reply]
That would be great. And, if I can make suggestions, that would be nice to have also available (not mandatory) descriptions linked to a path, so user can click on button called "Internal SD" instead of "/storage/emulated/0" since  "storage" and "emulated" are not readable (at least in my device) and if I click on them I'm kicked away to current folder.
But that's just a suggestion :-)

**Reply by aiekick** [reply]
you already have that via the bookmark pane, if you activate it

**Reply by aiekick** [reply]
do you have somewhere a imgui android project sample ?

**Reply by jenergy** [reply]
If you mean an android working apllication, I planned to release the first stable version before the end of September.
If you mean sources (with yours included, with very little changes), I planned to release all my project with sources at the end of October / November, some stuff is still missing, I'm working on the Nintendo Switch port too and I'd like to release ALL the sources (as clean as possible).
That can be ok for you?

**Reply by aiekick** [reply]
yep sure thanks. my goal is just to make changes for be compatible to android, but i can wait np

**Comment by ShawnWicky** [comment]
Hi,
I'm coding a filedialog to load obj files using your library. But i met a problem that i can't close the filedialog after clicking the "Ok" and "Close“ button. I copied your example code to construct a simple filedialog without custom font. The code is looks like the pic below. 
![code](https://user-images.githubusercontent.com/64520407/156528975-bdb308b3-9169-4352-b6a7-ebdce0a17617.png)
When i try to print the filepathname by clicking "OK" button on console, it works. But the filedialog does not close. I don't understand why. Casue i saw the API doc said the isOK() is responsible for closing the dialog, but I have it :(

**Reply by ShawnWicky** [reply]
when i click the "ok" or "close" it looks like it closed. But rendered the dialog again. I think it's a imgui or window render thing i got it wrong not the filedialog problem :(

**Reply by aiekick** [reply]
You have the code to open and display thz dialog in the same place.

So when the dialog is closed the open is called again each frame
 So its normal the dialog open again when you close it

The opendialog must be called only when needed by ex with a button
And the display must be called each framr.

Imgui is.not like classic gui, all things are computed each framr.

Check again the sample app and you will see that

**Reply by aiekick** [reply]
You found the solution ?

**Reply by ShawnWicky** [reply]
> You found the solution ?

Yes, thank you for the suggestion. I put the open somewhere else and it works perfectly.

**Comment by ShawnWicky** [comment]
Hi aiekick, I am trying to add a file dialog that locked the path it can choose. For example, I want the file dialog only to access the assets folder in my game engine. So can I set the file dialog access directory to a specific path and locked the user in this path in Iimguifiledialog?

**Reply by aiekick** [reply]
cool, a new use case.

not possible for the moment, but its an interesting feature to work on it.

question you want lock a root path, so the user can just select a file or also can explore the child directories ?

**Reply by ShawnWicky** [reply]
just select the file is ok I think

**Reply by aiekick** [reply]
ok ,will try

**Comment by Dean0809** [comment]
Hi aiekick, I met a bug that might relate to the path edition in the filedialog. When I click 'E', my program will shut down and pop out the assertion. Is that a problem from the imguifiledialog or my program
<img width="584" alt="Screen_Shot_2022-04-01_at_20 33 11" src="https://user-images.githubusercontent.com/50688122/161334019-ca063e16-0b4f-4a0b-ada4-e5e404497aba.png">
<img width="1111" alt="Screen_Shot_2022-04-01_at_20 32 46" src="https://user-images.githubusercontent.com/50688122/161334042-302b902c-11f1-42d2-880e-778f9461c781.png">

?

**Reply by aiekick** [reply]
yes you need to update imgui to the latest version.
the behavior for key have changed, you can check the sample app of ImGuiFileDialog

**Reply by Dean0809** [reply]
Thank you, it works when I update to the v0.6.4 of filedialog. However, I got a new bug in the source file. As shown in the picture. It looks like the USE_BOOKMARK in the header file is not working?
![image](https://user-images.githubusercontent.com/50688122/161378894-d24c9249-ba98-4543-bbb9-5c056c9fb752.png)

**Reply by aiekick** [reply]
yes you need to enable it, its a define. you can do that in a config file. check the doc for the feature bookmark

**Reply by aiekick** [reply]
its ok now ?

**Reply by Dean0809** [reply]
Yes, it works. Thank you very much! I am very sorry for the delay in replying to you

**Comment by paolotherocker** [comment]
Hi, I'm having trouble using the Directory selection mode. I can select a directory in the current path, but double clicking a directory doesn't open it.
Apologies if I'm missing something obvious, this is the code:
```cpp
    if (ImGui::Button("Open"))
    {
        ImGuiFileDialog::Instance()->OpenDialog("OpenBagDialog", "Open bag", nullptr, ".");
    }

    constexpr ImGuiWindowFlags file_dialog_flags =
        ImGuiWindowFlags_NoCollapse | ImGuiWindowFlags_NoSavedSettings | ImGuiWindowFlags_NoDocking;

    if (ImGuiFileDialog::Instance()->Display("OpenBagDialog", file_dialog_flags))
    {
        if (ImGuiFileDialog::Instance()->IsOk())
        {
            set_path(ImGuiFileDialog::Instance()->GetFilePathName());
        }

        ImGuiFileDialog::Instance()->Close();
    }
```
I'm building from the tip of the Lib_Only branch

**Reply by aiekick** [reply]
Seems good.
Can you try the test sample app and confirm than you have the same behavior. Maybe its a bug

**Reply by paolotherocker** [reply]
I can confirm it works ok with the test sample app. The only difference I can see is that I'm using SDL2 instead of GLFW.

**Reply by paolotherocker** [reply]
I also noticed that in my application, in the file picking mode, I only have to click once to enter a folder, where in the demo app I have to click twice to enter a folder

**Reply by aiekick** [reply]
ho its weird, ImGuiFileDialog is agnostic regarding the backend.
are you sure you are using the good keys ?
Can be an issue with the time delay for have a double click..
you are using the backened for SDL2 from ImGui ?
what si the version of ImGui you are using ? you know the key behavior have changed.

**Reply by paolotherocker** [reply]
> are you sure you are using the good keys ?

I'm not quite sure what that means

> you are using the backened for SDL2 from ImGui ?

Yes, I'm using SDL and OpenGL3, I'm following the example provided in the ImGui repository (example_sdl_opengl3)

> what si the version of ImGui you are using ? you know the key behavior have changed.

I'm building from the tip of the `docking` branch (commit `505f19e3`)

**Reply by paolotherocker** [reply]
I think I found the bug. You can reproduce the same behaviour in the sample app by uncommenting line 232:
```cpp
io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard;  // Enable Keyboard Controls
```

**Reply by aiekick** [reply]
the ImGuiConfigFlags_NavEnableKeyboard mode ofr ImGuiFileDialog is not wroking as expected since a lot of time. so pay attention with it.
maybe one i will succeed to do what i do but its a noodles plate.
i have applyed a little fix for get back the double click when you are using this mode.
for the other times, as explained on top of this thread, when you have a bug open an issue. this thread is just for general help

**Comment by Simeone0647** [comment]
Hi. I downloaded the FileDialog a couple of days ago, it's very cool and works fine, but is there a way for get the names of the selected files when you do multiselect? The file name is replaced with the n files selected text. I hope you can help me.

**Reply by aiekick** [reply]
do you checked the sample app ??

**Reply by Simeone0647** [reply]
Omg... Thanks, and sorry for the dumb question.

**Comment by Barashuk** [comment]
Hello. When used in my imgui window, Cyrillic characters are incorrectly displayed through ImGuiFileDialog. The most interesting thing is that everything is displayed normally in the test application. I added a font with Cyrillic support, both in the test application and in mine. The code that I use for my display. Use c++17 in both project with std::filesystem.
`auto& io = ImGui::GetIO();
ImGui::SetNextWindowPos(ImVec2(io.DisplaySize.x * 0.5f, io.DisplaySize.y * 0.5f), ImGuiCond_FirstUseEver, ImVec2(0.5f, 0.5f));
ImGui::SetNextWindowSize({ 800, 600 }, ImGuiCond_FirstUseEver);
ImGuiWindowFlags flags = ImGuiWindowFlags_NoCollapse | ImGuiWindowFlags_NoResize | ImGuiWindowFlags_NoMove | ImGuiWindowFlags_NoSavedSettings;
if (ImGui::Begin("Kill Informer", &bOpenMenu, flags)) {
	ImGui::Text(u8"Кирилица");
	if (ImGui::Button("test"))
	{
		fileDialog.OpenDialog("fff", "ffffff", ".*", "", 1, nullptr, ImGuiFileDialogFlags_NoDialog);
	}
	fileDialog.Display("fff", 32, { 0, 0 }, { 500, 400 });
	ImGui::End();
}`
![gta_sa_6wAWFvOH6d](https://user-images.githubusercontent.com/8094947/184867924-6d3f6885-849c-4451-b88b-d3b3f16303dc.png)

**Comment by aiekick** [comment]
Hello,

do you activated locales in your app ?

by ex : "std::setlocale(LC_ALL, ".UTF8");" at line 177 in main.cpp of the sample app

**Reply by Barashuk** [reply]
Thanks for help. Working <3

**Reply by aiekick** [reply]
to note, on windows, setlocale seems to work with msvc but not with mingw32, and maybe clang.
on linux,  there is no issues, not necessary to activate locales.. on macos i dont know, but maybe like for linux.
as usual windows is the bad child...

**Comment by damian-kos** [comment]
How do I make it show drives? Tried to build a sample app but struggled for couple hours as I don't know CMake yet so I need to figure it out. 
![image](https://github.com/aiekick/ImGuiFileDialog/assets/106775028/2d3aedfe-65f8-4227-a436-91dd8cdcc1bc)
tried wtih 
`IGFD::FileManager file_manager`
`file_manager.GetDrives();` but not sure how to utilize it

**Comment by aiekick** [comment]
since the last release, you have the drives in the side pane if you enable the USE_PLACES_FEATURE

you can just read the doc, all is explained => https://github.com/aiekick/ImGuiFileDialog?tab=readme-ov-file#places-

**Comment by 413x1nkp** [comment]
Hey, uhh... This might be a silly question, but how am I supposed to link this module if I don't plan on using `stdc++` ?
My intent was to use pure C, so I interact with Dear ImGui through cimgui bindings and so on. I thought this was a header-only library, as the README mentions that there are no plans to support compilation into `.so` files, however the `cpp` source file is still required for linking... What do I do though? I don't want to link whole `stdc++` and compilation as a library for this module isn't supported... The README mentions that C API is exposed, but... I can't really use it.

I suppose I have to compile this module alongside cimgui itself? Please let me know if that makes sense.

Best,
Alex

**Reply by aiekick** [reply]
Hello, ho i need to update the readme. since im using IGFD as a shared lib for many of my softs. just the CMakeList.txt not support that, but you can use your. see by example the shared used here : https://github.com/aiekick/ImGuiPack

**Reply by 413x1nkp** [reply]
> Hello, ho i need to update the readme. since im using IGFD as a shared lib for many of my softs. just the CMakeList.txt not support that, but you can use your. see by example the shared used here : [aiekick/ImGuiPack](https://github.com/aiekick/ImGuiPack)

While we're at it, I also don't think `cstdint` should be included unless `__cplusplus` is defined.
```c
/* lines 173-193 in ImGuiFileDialog.h */
#ifdef USE_THUMBNAILS
struct IGFD_Thumbnail_Info {
    int isReadyToDisplay            = 0;  // ready to be rendered, so texture created
    int isReadyToUpload             = 0;  // ready to upload to gpu
    int isLoadingOrLoaded           = 0;  // was sent to laoding or loaded
    int textureWidth                = 0;  // width of the texture to upload
    int textureHeight               = 0;  // height of the texture to upload
    int textureChannels             = 0;  // count channels of the texture to upload
    unsigned char* textureFileDatas = 0;  // file texture datas, will be rested to null after gpu upload
    void* textureID                 = 0;  // 2d texture id (void* is like ImtextureID type) (GL, DX, VK, Etc..)
    void* userDatas                 = 0;  // user datas
};
#endif  // USE_THUMBNAILS

#include <cstdint> /* <-- this header doesn't exist for C */

#ifdef __cplusplus

#ifndef IMGUI_DEFINE_MATH_OPERATORS
#define IMGUI_DEFINE_MATH_OPERATORS
#endif  // IMGUI_DEFINE_MATH_OPERATORS
```

**Reply by aiekick** [reply]
indeed you are right.  🤣

**Comment by EhadBalayla** [comment]
is there a way to set the root directory of the dialoge? because i want it to be in the assets folder in my engine so i can choose which directory to save object to

**Reply by aiekick** [reply]
please check the doc or the demo app

**Comment by saklis** [comment]
For the life of me, I can't include File Dialog into my project under MSVC.

The tricky part is that I'm using ImGui-SFML, that is compiling ImGui statically:


```CMake
set(IMGUI_SFML_FIND_SFML OFF)
set(IMGUI_DIR "${CMAKE_CURRENT_SOURCE_DIR}/3rd_party/DearImGui")
set(IMGUI_SFML_DIR "${CMAKE_CURRENT_SOURCE_DIR}/3rd_party/imgui-sfml")
set(IMGUI_SFML_DEPS sfml-system sfml-window sfml-graphics)
add_subdirectory(${IMGUI_SFML_DIR})
target_link_libraries(ImGui-SFML PRIVATE ${IMGUI_SFML_DEPS})
```

I compile and link File Dialog from that:

```CMake
set(IMGUIFILEDIALOG_DIR "${CMAKE_CURRENT_SOURCE_DIR}/3rd_party/ImGuiFileDialog")
file(GLOB IMGUIFILEDIALOG_SRC "${IMGUIFILEDIALOG_DIR}/ImGuiFileDialog.cpp")
add_library(ImGuiFileDialog STATIC ${IMGUIFILEDIALOG_SRC})
target_include_directories(ImGuiFileDialog PUBLIC ${IMGUIFILEDIALOG_DIR} ${IMGUI_DIR})
target_link_libraries(ImGuiFileDialog PUBLIC ImGui-SFML)
```

And this works perfectly on MinGW (Windows and Linux), but causes linker error on MSVC (VS22):

Error	LNK2001	unresolved external symbol "struct ImGuiContext * GImGui" (?GImGui@@3PEAUImGuiContext@@EA)	C:\cpp\aaa_projects\LOW-Engine\out\build\x64-Debug\LOW-Engine	C:\cpp\aaa_projects\LOW-Engine\out\build\x64-Debug\ImGuiFileDialog.lib(ImGuiFileDialog.cpp.obj)	1		

I'm somehwat new to C++ and the build system is driving me up the tree, so it's perfectly possible that it's something trivial I just don't know about.

The full CMakeLists is available here: [https://github.com/saklis/LOW-Engine/blob/main/CMakeLists.txt](https://github.com/saklis/LOW-Engine/blob/main/CMakeLists.txt)

**Reply by aiekick** [reply]
Hello dear, 

Since imgui is embedded in ImGui-SFML you need to link ImGui-SFML to ImGuiFileDialog. but it seems you have ImGui-SFML as SHARED lib and ImGuiFileDialog as STATIC library.
a shared-lib can be linked with static libs but not the inverse

So ImguiFileDialog must be shared

and you need after to define the declexport of the define 'IGD_API' like ImGui-SFLM do with 'IMGUI_SFML_API'

so maybe create a wrapper .h for call imguifiledialog.h like ImGui-SFLM with imgui-SFML_export.h

you can also frok ImGui-SFLM and embbed ImguiFileDialog  in it

check what i do with my shared lib ImGuipack : https://github.com/aiekick/ImGuiPack/blob/master/imguipack.h

Good Courage.

**Reply by saklis** [reply]
Kudos!

Now, trying to replicate imgui-SFML_export.h was over my head... But I followed your example with ImGuiPack and created one for myself (although, much simpler than yours).
Now my project compiles under MSVC, too.

Thank you for your help!

**Comment by Orofil** [comment]
Is it possible to display another modal on top of the file dialog modal? I'm trying to get a custom file overwrite alert to show up if the selected folder contains any of the files in a given set, but to display the modal it would have to be opened before the ImGui::EndPopup() of the file dialog modal, which is already called in ImGuiFileDialog::Instance()->Display().

This is basically what I'm trying to do:

``` C++
static std::string selected_path;
// Save window
if (ImGuiFileDialog::Instance()->Display("Save", ImGuiWindowFlags_Modal | ImGuiWindowFlags_NoCollapse))
{
    if (ImGuiFileDialog::Instance()->IsOk())
    {
        selected_path = ImGuiFileDialog::Instance()->GetCurrentPath();
        // If any of the config files exist, show a warning
        std::set<std::string> config_files = {"file1.json", "file2.json", "folder/file3.json"};
        bool show_warning = false;
        for (const auto& file : config_files)
        {
            if (std::filesystem::exists(std::filesystem::path(selected_path) / file))
            {
                show_warning = true;
            }
        }
        if (show_warning)
        {
            ImGui::OpenPopup("Overwrite warning");
        }
        else
        {
            saveFiles(selected_path);
            ImGuiFileDialog::Instance()->Close();
        }
    }
    else
    {
        ImGuiFileDialog::Instance()->Close();
    }
}

// Overwrite warning window
if (ImGui::BeginPopupModal("Overwrite warning", nullptr, ImGuiWindowFlags_AlwaysAutoResize)) {
    ImGui::Text("Saving will overwrite existing files. Continue?");
    if (ImGui::Button("OK"))
    {
        saveFiles(selected_path);
        ImGuiFileDialog::Instance()->Close();
        ImGui::CloseCurrentPopup();
    }
    ImGui::SameLine();
    if (ImGui::Button("Cancel"))
    {
        ImGui::CloseCurrentPopup();
    }
    ImGui::EndPopup();
}
```

But of course the overwrite warning doesn't show up this way. I thought a hacky way to make this work would be to set `m_FileDialogInternal.okResultToConfirm` to true so ImGui::EndPopup() doesn't get called, but it's protected so it's not really accessible, and probably not a good idea.

For now I'm doing this by closing the file dialog whenever the overwrite warning should be shown, and then re-opening it when clicking cancel. This loses any selections etc. though.
